File: sdks/ios/Sources/RajutechieStreamKit/Modules/StreamModule.swift
import Foundation

public struct LiveStream: Codable {
    public let id: String
    public let title: String
    public let streamKey: String
    public let status: String
    public let hlsUrl: String?
    public let viewerCount: Int
}

public final class StreamModule {
    private let http: HTTPClient
    private let ws: WebSocketClient

    init(http: HTTPClient, ws: WebSocketClient) {
        self.http = http
        self.ws = ws
    }

    public func create(title: String, visibility: String = "public") async throws -> LiveStream {
        struct CreateStreamBody: Encodable {
            let title: String
            let visibility: String
        }
        let result: LiveStream = try await http.post("/streams", body: CreateStreamBody(title: title, visibility: visibility))
        return result
    }

    public func get(_ streamId: String) async throws -> LiveStream {
        let result: LiveStream = try await http.get("/streams/\(streamId)")
        return result
    }

    public func start(_ streamId: String) async throws -> LiveStream {
        let result: LiveStream = try await http.post("/streams/\(streamId)/start")
        return result
    }

    public func stop(_ streamId: String) async throws -> LiveStream {
        let result: LiveStream = try await http.post("/streams/\(streamId)/stop")
        return result
    }

    public func getViewerCount(_ streamId: String) async throws -> Int {
        let result: ViewerCount = try await http.get("/streams/\(streamId)/viewers")
        return result.count
    }
}

private struct ViewerCount: Codable {
    let count: Int
}

File: sdks/android/streamkit-sdk/src/main/kotlin/com/rajutechie/streamkit/modules/StreamModule.kt
package com.rajutechie.streamkit.modules

import com.rajutechie.streamkit.transport.HttpClient
import com.rajutechie.streamkit.transport.WebSocketClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow

data class LiveStream(val id: String, val title: String, val streamKey: String, val status: String, val hlsUrl: String?, val viewerCount: Int)

class StreamModule(
    private val http: HttpClient,
    private val ws: WebSocketClient,
    private val scope: CoroutineScope
) {
    suspend fun create(title: String, visibility: String = "public"): LiveStream {
        return http.post("/streams", mapOf("title" to title, "visibility" to visibility))
    }

    suspend fun get(streamId: String): LiveStream = http.get("/streams/$streamId")
    suspend fun start(streamId: String): LiveStream = http.post("/streams/$streamId/start")
    suspend fun stop(streamId: String): LiveStream = http.post("/streams/$streamId/stop")

    suspend fun getViewerCount(streamId: String): Int {
        val result: Map<String, Any> = http.get("/streams/$streamId/viewers")
        return (result["count"] as? Double)?.toInt() ?: 0
    }

    fun observeViewerCount(): Flow<Map<String, Any>> = ws.on("stream.viewer.count", Map::class.java) as Flow<Map<String, Any>>
}
